pallets/settlement/src/lib.rs

```diff
@@ -1609,101 +1644,124 @@ impl<T: Config> Module<T> {
+        ));
 
-        match frame_storage_with_transaction(|| {
-            Self::release_asset_locks_and_transfer_pending_legs(instruction_id, &instruction_legs)
+        // Verifies if the venue is allowed for all tickers in the instruction
+        Self::ensure_allowed_venue(&instruction_legs, venue_id, weight_meter)?;
+
+        // Attempts to release the locks and transfer all fungible an non fungible assets
+        if let Err(leg_id) = frame_storage_with_transaction(|| {
+            Self::release_asset_locks_and_transfer_pending_legs(
+                instruction_id,
+                &instruction_legs,
+                weight_meter,
+            )
         })? {
-            Ok(_) => {
-                Self::deposit_event(RawEvent::InstructionExecuted(
-                    SettlementDID.as_id(),
-                    instruction_id,
-                ));
-            }
-            Err(leg_id) => {
-                Self::deposit_event(RawEvent::LegFailedExecution(
-                    SettlementDID.as_id(),
-                    instruction_id,
-                    leg_id,
-                ));
-                Self::deposit_event(RawEvent::InstructionFailed(
-                    SettlementDID.as_id(),
-                    instruction_id,
-                ));
-                // Unclaim receipts for the failed transaction so that they can be reused
-                Self::unsafe_unclaim_receipts(instruction_id, &instruction_legs);
-                return Err(Error::<T>::InstructionFailed.into());
-            }
+            Self::deposit_event(RawEvent::LegFailedExecution(
+                SettlementDID.as_id(),
+                instruction_id,
+                leg_id,
+            ));
+            Self::deposit_event(RawEvent::InstructionFailed(
+                SettlementDID.as_id(),
+                instruction_id,
+            ));
+            // Unclaim receipts for the failed transaction so that they can be reused
+            Self::unsafe_unclaim_receipts(instruction_id, &instruction_legs);
+            return Err(Error::<T>::InstructionFailed.into());
         }
 
-        Ok(instruction_legs.len().try_into().unwrap_or_default())
+        Self::deposit_event(RawEvent::InstructionExecuted(
+            SettlementDID.as_id(),
+            instruction_id,
+        ));
+        Ok(())
     }
 
     fn release_asset_locks_and_transfer_pending_legs(
         instruction_id: InstructionId,
         instruction_legs: &[(LegId, LegV2)],
+        weight_meter: &mut WeightMeter,
     ) -> TransactionOutcome<Result<Result<(), LegId>, DispatchError>> {
         Self::unchecked_release_locks(instruction_id, instruction_legs);
+        // Charges the weight consumed
+        let transfer_data = Self::get_transfer_data(&instruction_legs).unwrap();
+        weight_meter.check_accrue(<T as Config>::WeightInfo::unchecked_release_locks(
+            transfer_data.fungible(),
+            transfer_data.non_fungible(),
+        ));
```

Comment:
The `check_accrue` method will not throw an error, it just returns a `bool`.  Also if the weight would put it over the limit, it doesn't update the consumed weight counter.

Looks like we will need to wrap `WeightMeter` in our own "meter" object and have a `charge` method that throws an error.

Reply:
Also we need to try charging the meter before the action.